apps/deploy_monitoring/src/monitors/resource.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) SimpleStaking and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+#![forbid(unsafe_code)]
+
+use crate::monitors::Alerts;
+use crate::node::OcamlNode;
+use crate::slack::SlackServer;
+use chrono::Utc;
+
+use serde::Serialize;
+use slog::Logger;
+use std::collections::VecDeque;
+use std::sync::{Arc, RwLock};
+
+use getset::Getters;
+use sysinfo::{System, SystemExt};
+
+use shell::stats::memory::ProcessMemoryStats;
+
+use crate::display_info::DiskData;
+use crate::node::{Node, TezedgeNode, OCAML_PORT, TEZEDGE_PORT};
+
+pub type ResourceUtilizationStorage = Arc<RwLock<VecDeque<ResourceUtilization>>>;
+
+/// The max capacity of the VecDeque holding the measurements
+pub const MEASUREMENTS_MAX_CAPACITY: usize = 40320;
+
+pub struct ResourceMonitor {
+    ocaml_resource_utilization: ResourceUtilizationStorage,
+    tezedge_resource_utilization: ResourceUtilizationStorage,
+    last_checked_head_level: Option<u64>,
+    alerts: Alerts,
+    log: Logger,
+    slack: SlackServer,
+    system: System,
+}
+
+#[derive(Clone, Debug, Default, Serialize, Getters)]
+pub struct MemoryStats {
+    #[get = "pub(crate)"]
+    node: ProcessMemoryStats,
+
+    #[get = "pub(crate)"]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    protocol_runners: Option<Vec<ProcessMemoryStats>>,
+
+    #[get = "pub(crate)"]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    validators: Option<Vec<ProcessMemoryStats>>,
+}
+
+#[derive(Clone, Debug, Serialize, Getters)]
+pub struct ResourceUtilization {
+    #[get = "pub(crate)"]
+    timestamp: i64,
+
+    #[get = "pub(crate)"]
+    memory: MemoryStats,
+
+    #[get = "pub(crate)"]
+    disk: DiskData,
+
+    #[get = "pub(crate)"]
+    cpu: CpuStats,
+}
+
+#[derive(Clone, Debug, Serialize, Getters)]
+pub struct CpuStats {
+    #[get = "pub(crate)"]
+    node: i32,
+
+    #[get = "pub(crate)"]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    protocol_runners: Option<i32>,
+}
+
+impl ResourceMonitor {
+    pub fn new(
+        ocaml_resource_utilization: ResourceUtilizationStorage,
+        tezedge_resource_utilization: ResourceUtilizationStorage,
+        last_checked_head_level: Option<u64>,
+        alerts: Alerts,
+        log: Logger,
+        slack: SlackServer,
+    ) -> Self {
+        Self {
+            ocaml_resource_utilization,
+            tezedge_resource_utilization,
+            last_checked_head_level,
+            alerts,
+            log,
+            slack,
+            system: System::new_all(),
+        }
+    }
+
+    pub async fn take_measurement(&mut self) -> Result<(), failure::Error> {
```

Comment:
@adonagy 
I would not mix measurements,
if you refactor it to `resource_utilization: Vec<ResourceUtilizationStorage>`,
then you could iterate it and `take_measurement` for every entry